src/software/new_geom/linear_spline_test.cpp

```diff
@@ -0,0 +1,197 @@
+    EXPECT_TRUE(s.getValueAt(0.5).isClose(Point(0.5, 5.5), 1e-9));
+    EXPECT_TRUE(s.getValueAt(0.6).isClose(Point(1.2, 10.8), 1e-7));
+    EXPECT_TRUE(s.getValueAt(0.9).isClose(Point(3.3, 26.7), 1e-7));
+    EXPECT_EQ(s.getValueAt(1.0), points[1]);
+}
+
+TEST(LinearSplineTest, test_three_knot_value_at_out_of_range)
+{
+    std::vector<Point> points({Point(4, 18), Point(3, 0), Point(-7, -1)});
+    LinearSpline2d s(points);
+
+    bool failing = true;
+
+    try
+    {
+        s.getValueAt(-0.1);
+    }
+    catch (std::invalid_argument &e)
+    {
+        failing = false;
+    }
+
+    if (failing)
+    {
+        ADD_FAILURE()
+            << "LinearSpline2d returning value for value (-0.1) outside of domain";
+        return;
+    }
+
+    try
+    {
+        s.getValueAt(1.1);
+    }
+    catch (std::invalid_argument &e)
+    {
+        failing = false;
+    }
+
+    if (failing)
+    {
+        ADD_FAILURE()
+            << "LinearSpline2d returning value for value (1.1) outside of domain";
+        return;
+    }
+}
+
+TEST(LinearSplineTest, test_one_knot_value_at_out_of_range)
+{
+    std::vector<Point> points({Point(1, -18)});
+    LinearSpline2d s(points);
+
+    bool failing = true;
+
+    try
+    {
+        s.getValueAt(-0.1);
+    }
+    catch (std::invalid_argument &e)
+    {
+        failing = false;
+    }
+
+    if (failing)
+    {
+        ADD_FAILURE()
+            << "LinearSpline2d returning value for value (-0.1) outside of domain";
+        return;
+    }
+
+    try
+    {
+        s.getValueAt(1.1);
+    }
+    catch (std::invalid_argument &e)
+    {
+        failing = false;
+    }
+
+    if (failing)
+    {
+        ADD_FAILURE()
+            << "LinearSpline2d returning value for value (1.1) outside of domain";
+        return;
+    }
+}
+
+TEST(LinearSplineTest, test_two_knot_value_at_out_of_range)
+{
+    std::vector<Point> points({Point(1, -2), Point(1, -18)});
+    LinearSpline2d s(points);
+
+    bool failing = true;
+
+    try
+    {
+        s.getValueAt(-0.1);
+    }
+    catch (std::invalid_argument &e)
+    {
+        failing = false;
+    }
+
+    if (failing)
+    {
+        ADD_FAILURE()
+            << "LinearSpline2d returning value for value (-0.1) outside of domain";
+        return;
+    }
+
+    try
+    {
+        s.getValueAt(1.1);
+    }
+    catch (std::invalid_argument &e)
+    {
+        failing = false;
+    }
+
+    if (failing)
+    {
+        ADD_FAILURE()
+            << "LinearSpline2d returning value for value (1.1) outside of domain";
+        return;
+    }
+}
+
+TEST(LinearSplineTest, test_spline_points_list_constructor)
+{
+    std::vector<Point> points({Point(1, 2), Point(2, 3), Point(0, -1)});
+    LinearSpline2d s({Point(1, 2), Point(2, 3), Point(0, -1)});
+    EXPECT_EQ(s.getNumKnots(), points.size());
+    std::vector<Point> spline_points = s.getKnots();
+    EXPECT_EQ(spline_points, points);
+    EXPECT_EQ(s.getStartPoint(), points[0]);
```

Comment:
Consider separate unit tests for the start and end point functions (and numKnots as per my above comment)